src/sp800_108.rs

```diff
@@ -0,0 +1,396 @@
+// Copyright 2024 Simo Sorce
+// See LICENSE.txt file for terms
+
+fn key_to_segment_size(key: usize, segment: usize) -> usize {
+    ((key + segment - 1) / segment) * segment
+}
+
+const MAX8: usize = u8::MAX as usize;
+const MAX16: usize = u16::MAX as usize;
+const MAX24: usize = MAX8 * MAX16;
+const MAX32: usize = u32::MAX as usize;
+const MAX40: usize = MAX8 * MAX32;
+const MAX48: usize = MAX16 * MAX32;
+const MAX56: usize = MAX8 * MAX16 * MAX32;
+const MAX64: usize = u64::MAX as usize;
+
+fn ctr_update(
+    param: &Sp800CounterFormat,
+    ctr: usize,
+    op: &mut Box<dyn Sign>,
+) -> KResult<()> {
+    if !param.defined {
+        return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+    }
+    match param.bits {
+        8 => {
+            if ctr > MAX8 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+            op.sign_update(&[ctr as u8])
+        }
+        16 => {
+            if ctr > MAX16 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+            let data = if param.le {
+                (ctr as u16).to_le_bytes()
+            } else {
+                (ctr as u16).to_be_bytes()
+            };
+            op.sign_update(&data)
+        }
+        24 => {
+            if ctr > MAX24 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+            let (data, s, e) = if param.le {
+                ((ctr as u32).to_le_bytes(), 0, 3)
+            } else {
+                ((ctr as u32).to_be_bytes(), 1, 4)
+            };
+            op.sign_update(&data[s..e])
+        }
+        32 => {
+            if ctr > MAX32 {
+                return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+            }
+            let data = if param.le {
+                (ctr as u32).to_le_bytes()
+            } else {
+                (ctr as u32).to_be_bytes()
+            };
+            op.sign_update(&data)
+        }
+        _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+    }
+}
+
+fn dkm_update(
+    param: &Sp800DKMLengthFormat,
+    klen: usize,
+    slen: usize,
+    op: &mut Box<dyn Sign>,
+) -> KResult<()> {
+    let len = match param.method {
+        CK_SP800_108_DKM_LENGTH_SUM_OF_SEGMENTS => slen,
+        CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS => klen,
+        _ => return err_rv!(CKR_MECHANISM_PARAM_INVALID),
+    };
+    /* up to 64 bits */
+    match param.bits {
+        8 => {
+            if len > MAX8 {
```

Comment:
I believe len is allowed to be > MAX8. you need to take the low 8 bits in that case. Same with 16, 24, and 32, so you just need to mask off the len before you convert it to data (len & 0xff for 8, len & 0xffff for 15, etc).

Reply:
I have a patch to change these errors into just truncating the value, will submit soon, thanks for spotting this.